doc/man/doveadm-mount.1.in
.\" Copyright (c) 2012-2018 Dovecot authors, see the included COPYING file
.TH DOVEADM\-MOUNT 1 "2012-02-16" "Dovecot v2.2" "Dovecot"
.SH NAME
doveadm\-mount \- Manage the list of mountpoints where mails are stored
.\"------------------------------------------------------------------------
.SH SYNOPSIS
.BR doveadm " [" \-Dv "] [" \-f
.IR formatter ]
.B mount
.IR command " [" arguments ]
.\"------------------------------------------------------------------------
.SH DESCRIPTION
The doveadm
.B mount
.IR command s
can be used to manage the list of mountpoints where mails are stored.
This is used mainly for better error handling when a mountpoint isn\(aqt
mounted for some reason:
.TP 4
*
If a mail directory doesn\(aqt exist, it\(aqs autocreated.
If the user\(aqs mails aren\(aqt mounted and filesystem permissions still
allow the autocreation, the user will see an empty mailbox and later will
have to redownload all mails.
If the mountpoint is known to be unmounted, Dovecot will simply fail
opening any mailboxes.
.TP
*
If dbox alternate storage isn\(aqt mounted and a mail in it is attempted to
be accessed, Dovecot normally rebuilds the indexes and notices that all the
mails in alt storage are expunged.
When the alt storage is mounted back and even if index is again rebuilt,
the mails won\(aqt necessarily become visible anymore for IMAP clients.
If the mountpoint is known to be unmounted, Dovecot won\(aqt rebuild
indexes and lose the mails.
.PP
Dovecot automatically adds mountpoints to this list at startup.
If you don\(aqt want some of the mountpoints added, you can add a wildcard
ignore for it.
.\"------------------------------------------------------------------------
@INCLUDE:global-options-formatter@
.\"------------------------------------------------------------------------
.SH ARGUMENTS
.TP
.I path
The directory name of a mountpoint
.\"-------------------------------------
.TP
.I state
The
.I state
of a mountpoint.
Either
.BR online " or " ignore .
.\"------------------------------------------------------------------------
.SH COMMANDS
.SS mount add
.B doveadm mount add
.RI [ path " [" state ]]
.PP
If this command is run without any parameters, doveadm detects all missing
mountpoints and adds them (the same way as when Dovecot does at startup).
.PP
When a mountpoint
.I path
is given, it\(aqs added as a mountpoint.
The
.I state
can currently be either
.RB \(dq online \(dq
(default) or
.RB \(dq ignore \(dq.
The ignore state is mainly useful with path wildcards to add mountpoints
that you never want Dovecot to automatically add, such as:
.B  doveadm mount add '/mnt/*' ignore
.\"-------------------------------------
.SS mount list
.BR doveadm " [" \-Dv " ] [" \-f
.IR formatter ]
.B mount list
.PP
This command lists the mountpoints known to Dovecot and their state.
.\"-------------------------------------
.SS mount remove
.BI "doveadm mount remove " path
.PP
This command removes the specified
.I path
from the mountpoint list.
.\"------------------------------------------------------------------------
@INCLUDE:reporting-bugs@
.\"------------------------------------------------------------------------
.SH SEE ALSO
.BR doveadm (1)

NEWS
v2.3.0 2017-12-22  Timo Sirainen <tss@iki.fi>

	* Various setting changes, see https://wiki2.dovecot.org/Upgrading/2.3
	* Logging rewrite started: Logging is now based on hierarchical events.
	  This makes it possible to do various things, like: 1) giving
	  consistent log prefixes, 2) enabling debug logging with finer
	  granularity, 3) provide logs in more machine readable formats
	  (e.g. json). Everything isn't finished yet, especially a lot of the
	  old logging code still needs to be translated to the new way.
	* Statistics rewrite started: Stats are now based on (log) events.
	  It's possible to gather statistics about any event that is logged.
	  See http://wiki2.dovecot.org/Statistics for details
	* ssl_dh setting replaces the old generated ssl-parameters.dat
	* IMAP: When BINARY FETCH finds a broken mails, send [PARSE] error
	  instead of [UNKNOWNCTE]
	* Linux: core dumping via PR_SET_DUMPABLE is no longer enabled by
	  default due to potential security reasons (found by cPanel Security
	  Team).

	+ Added support for SMTP submission proxy server, which includes
	  support for BURL and CHUNKING extension.
	+ LMTP rewrite. Supports now CHUNKING extension and mixing of
	  local/proxy recipients.
	+ auth: Support libsodium to add support for ARGON2I and ARGON2ID
	  password schemes.
	+ auth: Support BLF-CRYPT password scheme in all platforms
	+ auth: Added LUA scripting support for passdb/userdb.
	  See https://wiki2.dovecot.org/AuthDatabase/Lua
	- Input streams are more reliable now when there are errors or when
	  the maximum buffer size is reached. Previously in some situations
	  this could have caused Dovecot to try to read already freed memory.
	- Output streams weren't previously handling failures when writing a
	  trailer at the end of the stream. This mainly affected encrypt and
	  zlib compress ostreams, which could have silently written truncated
	  files if the last write happened to fail (which shouldn't normally
	  have ever happened).
	- virtual plugin: Fixed panic when fetching mails from virtual
	  mailboxes with IMAP BINARY extension.
	- doveadm-server: Fix potential hangs with SSL connections
	- doveadm proxy: Reading commands' output from v2.2.33+ servers could
	  have caused the output to be corrupted or caused a crash.
	- Many other smaller fixes

v2.2.33.2 2017-10-20  Timo Sirainen <tss@iki.fi>

	- doveadm: Fix crash in proxying (or dsync replication) if remote is
	  running older than v2.2.33
	- auth: Fix memory leak in %{ldap_dn}
	- dict-sql: Fix data types to work correctly with Cassandra

v2.2.33.1 2017-10-10  Timo Sirainen <tss@iki.fi>

	- dovecot-lda was logging to stderr instead of to the log file.

v2.2.33 2017-10-10  Timo Sirainen <tss@iki.fi>

	* doveadm director commands wait for the changes to be visible in the
	  whole ring before they return. This is especially useful in testing.
	* Environments listed in import_environment setting are now set or
	  preserved when executing standalone commands (e.g. doveadm)

	+ doveadm proxy: Support proxying logs. Previously the logs were
	  visible only in the backend's logs.
	+ Added %{if}, see https://wiki2.dovecot.org/Variables#Conditionals
	+ Added a new notify_status plugin, which can be used to update dict
	  with current status of a mailbox when it changes. See
	  https://wiki2.dovecot.org/Plugins/NotifyStatus
	+ Mailbox list index can be disabled for a namespace by appending
	  ":LISTINDEX=" to location setting.
	+ dsync/imapc: Added dsync_hashed_headers setting to specify which
	  headers are used to match emails.
	+ pop3-migration: Add pop3_migration_ignore_extra_uidls=yes to ignore
	  mails that are visible in POP3 but not IMAP. This could happen if
	  new mails were delivered during the migration run.
	+ pop3-migration: Further improvements to help with Zimbra
	+ pop3-migration: Cache POP3 UIDLs in imapc's dovecot.index.cache
	  if indexes are enabled. These are used to optimize incremental syncs.
	+ cassandra, dict-sql: Use prepared statements if protocol version>3.
	+ auth: Added %{ldap_dn} variable for passdb/userdb ldap
	- acl: The "create" (k) permission in global acl-file was sometimes
	  ignored, allowing users to create mailboxes when they shouldn't have.
	- sdbox: Mails were always opened when expunging, unless
	  mail_attachment_fs was explicitly set to empty.
	- lmtp/doveadm proxy: hostip passdb field was ignored, which caused
	  unnecessary DNS lookups if host field wasn't an IP
	- lmtp proxy: Fix crash when receiving unexpected reply in RCPT TO
	- quota_clone: Update also when quota is unlimited (broken in v2.2.31)
	- mbox, zlib: Fix assert-crash when accessing compressed mbox
	- doveadm director kick -f parameter didn't work
	- doveadm director flush <host> resulted flushing all hosts, if <host>
	  wasn't an IP address.
	- director: Various fixes to handling backend/director changes at
	  abnormal times, especially while ring was unsynced. These could have
	  resulted in crashes, non-optimal behavior or ignoring some of the
	  changes.
	- director: Use less CPU in imap-login processes when moving/kicking
	  many users.
	- lmtp: Session IDs were duplicated/confusing with multiple RCPT TOs
	  when lmtp_rcpt_check_quota=yes
	- doveadm sync -1 fails when local mailboxes exist that do not exist
	  remotely. This commonly happened when lazy_expunge mailbox was
	  autocreated when incremental sync expunged mails.
	- pop3: rawlog_dir setting didn't work


v2.2.32 2017-08-24  Timo Sirainen <tss@iki.fi>

	* imapc: Info-level line is logged every time when successfully
	  connected to the remote server. This includes local/remote IP/port,
	  which can be useful for matching against external logs.
	* config: Log a warning if plugin { key=no } is used explicitly.
	  v2.3 will support "no" properly in plugin settings, but for now
	  any value at all for a boolean plugin setting is treated as "yes",
	  even if it's written as explicit "no". This change will now warn

src/doveadm/doveadm-mount.c
/* Copyright (c) 2013-2018 Dovecot authors, see the included COPYING file */

#include "lib.h"
#include "master-service-settings.h"
#include "mountpoint-list.h"
#include "doveadm.h"
#include "doveadm-print.h"

extern struct doveadm_cmd doveadm_cmd_mount[];

static void mount_cmd_help(doveadm_command_t *cmd) ATTR_NORETURN;

static struct mountpoint_list *mountpoint_list_get(void)
{
	const char *perm_path, *state_path;

	perm_path = t_strconcat(service_set->state_dir,
				 "/"MOUNTPOINT_LIST_FNAME, NULL);
	state_path = t_strconcat(service_set->base_dir,
				 "/"MOUNTPOINT_LIST_FNAME, NULL);
	return mountpoint_list_init(perm_path, state_path);
}

static void cmd_mount_list(int argc, char *argv[])
{
	struct mountpoint_list *mountpoints;
	struct mountpoint_list_iter *iter;
	struct mountpoint_list_rec *rec;
	bool mounts_known;

	if (argc > 2)
		mount_cmd_help(cmd_mount_list);

	doveadm_print_init(DOVEADM_PRINT_TYPE_TABLE);
	doveadm_print_header_simple(" ");
	doveadm_print_header("path", "path", DOVEADM_PRINT_HEADER_FLAG_EXPAND);
	doveadm_print_header_simple("state");

	mountpoints = mountpoint_list_get();
	mounts_known = mountpoint_list_update_mounted(mountpoints) == 0;
	iter = mountpoint_list_iter_init(mountpoints);
	while ((rec = mountpoint_list_iter_next(iter)) != NULL) {
		if (argv[1] != NULL && strcmp(argv[1], rec->mount_path) != 0)
			continue;

		if (mounts_known && MOUNTPOINT_WRONGLY_NOT_MOUNTED(rec))
			doveadm_print("!");
		else
			doveadm_print(" ");
		doveadm_print(!rec->wildcard ? rec->mount_path :
			      t_strconcat(rec->mount_path, "*", NULL));

		doveadm_print(rec->state);
	}
	mountpoint_list_iter_deinit(&iter);
	mountpoint_list_deinit(&mountpoints);
}

static bool mount_path_get_wildcard(const char **path)
{
	unsigned int len;

	len = strlen(*path);
	if (len > 0 && (*path)[len-1] == '*') {
		*path = t_strndup(*path, len-1);
		return TRUE;
	} else {
		return FALSE;
	}
}

static void cmd_mount_add(int argc, char *argv[])
{
	struct mountpoint_list *mountpoints;
	struct mountpoint_list_rec rec;
	int ret = 0;

	if (argc > 3)
		mount_cmd_help(cmd_mount_add);

	mountpoints = mountpoint_list_get();
	if (argv[1] == NULL) {
		ret = mountpoint_list_add_missing(mountpoints,
			MOUNTPOINT_STATE_DEFAULT,
			mountpoint_list_default_ignore_prefixes,
			mountpoint_list_default_ignore_types);
	} else {
		i_zero(&rec);
		rec.mount_path = argv[1];
		rec.state = argv[2] != NULL ? argv[2] :
			MOUNTPOINT_STATE_DEFAULT;

		if (mount_path_get_wildcard(&rec.mount_path))
			rec.wildcard = TRUE;
		mountpoint_list_add(mountpoints, &rec);
	}
	if (mountpoint_list_save(mountpoints) < 0)
		ret = -1;
	mountpoint_list_deinit(&mountpoints);
	if (ret < 0)
		doveadm_exit_code = EX_TEMPFAIL;
}

static void cmd_mount_remove(int argc, char *argv[])
{
	struct mountpoint_list *mountpoints;
	const char *mount_path;

	if (argc != 2)
		mount_cmd_help(cmd_mount_remove);

	mount_path = argv[1];
	(void)mount_path_get_wildcard(&mount_path);

	mountpoints = mountpoint_list_get();
	if (!mountpoint_list_remove(mountpoints, mount_path))
		i_error("Mountpoint not found: %s", mount_path);
	else
		(void)mountpoint_list_save(mountpoints);
	mountpoint_list_deinit(&mountpoints);
}

struct doveadm_cmd doveadm_cmd_mount[] = {
	{ cmd_mount_list, "mount list", "[<path>]" },
	{ cmd_mount_add, "mount add", "[<path> [<state>]]" },
	{ cmd_mount_remove, "mount remove", "<path>" }
};

static void mount_cmd_help(doveadm_command_t *cmd)
{
	unsigned int i;

	for (i = 0; i < N_ELEMENTS(doveadm_cmd_mount); i++) {
		if (doveadm_cmd_mount[i].cmd == cmd)
			help(&doveadm_cmd_mount[i]);
	}
	i_unreached();
}

void doveadm_register_mount_commands(void)
{
	unsigned int i;

	for (i = 0; i < N_ELEMENTS(doveadm_cmd_mount); i++)
		doveadm_register_cmd(&doveadm_cmd_mount[i]);
}
